plugins/destination/postgresql/client/migrate.go

```diff
@@ -48,165 +167,193 @@ func (c *Client) Migrate(ctx context.Context, tables schema.Tables) error {
+	columnName := pgx.Identifier{column.Name}.Sanitize()
+	columnType := c.SchemaTypeToPg(column.Type)
+	// try alter column in place first
+	sql := "alter table " + tableName + " alter column " + columnName + " type " + columnType
+	if column.CreationOptions.NotNull {
+		sql += " set not null"
 	}
-
-	if pgPKs, err = c.getPgTablePrimaryKeys(ctx, table.Name); err != nil {
-		return fmt.Errorf("failed to get table %s primary key columns: %w", table.Name, err)
+	if _, err := c.conn.Exec(ctx, sql); err != nil {
+		c.logger.Warn().Err(err).Str("table", tableName).Str("column", column.Name).Msg("Column type changed in place failed.")
 	}
-
-	stalePks := c.getStalePks(pgPKs, table)
-	constraintName := pgx.Identifier{table.Name + "_cqpk"}.Sanitize()
-	dropConstraintSQL := "alter table " + tableName + " drop constraint if exists " + constraintName
-	if len(stalePks) > 0 {
-		message := "the following primary keys were removed from the schema %q for table %q.\nYou can migrate the table manually by running:\n%s"
-		sep := strings.Repeat("-", len(dropConstraintSQL)+1)
-		query := fmt.Sprintf("%s\n%s;\n%s\n%s", sep, dropConstraintSQL, getDropNotNullQuery(table, stalePks), sep)
-		return fmt.Errorf(message, stalePks, table.Name, query)
+	tx, err := c.conn.BeginTx(ctx, pgx.TxOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to start alter column transaction: %w", err)
 	}
-
-	reCreatePrimaryKeys := false
-
-	for _, col := range table.Columns {
-		columnName := pgx.Identifier{col.Name}.Sanitize()
-		columnType := c.SchemaTypeToPg(col.Type)
-		if columnType == "" {
-			c.logger.Warn().Str("table", table.Name).Str("column", col.Name).Msg("Column type not supported, skipping")
-			continue
+	defer func() {
+		if err := tx.Rollback(ctx); err != nil {
+			c.logger.Error().Err(err).Msg("Failed to rollback alter column transaction")
 		}
-		pgColumn := pgColumns.getPgColumn(col.Name)
+	}()
 
-		switch {
-		case pgColumn == nil:
-			c.logger.Info().Str("table", table.Name).Str("column", col.Name).Msg("Column doesn't exist, creating")
+	sql = "alter table " + tableName + " drop column " + columnName
+	// right now we will drop the column and re-create. in the future we will have an option to automigrate
+	if _, err := tx.Exec(ctx, sql); err != nil {
+		return fmt.Errorf("failed to drop column %s.%s: %w", column.Name, tableName, err)
+	}
+	sql = "alter table " + tableName + " add column " + columnName + " " + columnType
+	if column.CreationOptions.NotNull {
+		sql += " not null"
+	}
+	if _, err := tx.Exec(ctx, sql); err != nil {
+		var pgErr *pgconn.PgError
+		if !errors.As(err, &pgErr) {
+			return fmt.Errorf("failed to add column %s.%s: %w", tableName, column.Name, err)
+		}
+		// this means the column contains null value and we cannot recreate it without knowing what is the default
+		// This is usually the case when the column is a primary key
+		// the user will have to drop and recreate the table
+		if pgErr.Code == "23502" {
```

Comment:
I don't understand this error report. If we drop and add a not null column on an existing table won't we always reach this code? I mean by dropiing the column we remove any existing values, so even if the column had all not null values we're basically removing them at the beginging of the transaction

Reply:
Indeed. maybe it's left overs. pretty strange it's here.